### PR TITLE
Support Filtering on Large List encoded by Bitmap

### LINKED ISSUE #12341: [Feature Request] Support more efficient encoding of filters over many terms
### Is your feature request related to a problem? Please describe

I recently read [this blog post](https://luis-sena.medium.com/improve-elasticsearch-filtering-performance-10x-using-this-plugin-8c6485516c1a), where the author claims a 10x speedup on a large `terms` query by encoding the field values as a roaring bitmap. 

I believe that part of the improvement comes from the use of doc values to post-filter hits that come from a lead iterator, which is now the OpenSearch (starting with 2.12) thanks to @harshavamsi's changes in https://github.com/opensearch-project/OpenSearch/pull/11209 to support IndexOrDocValuesQuery for all numeric query types. (Behind the scenes, Lucene implements the DV query using a `LongHashSet`, which I think should perform similarly to RoaringBitmap.)

The more interesting part (IMO) is that the roaring bitmap of numeric terms gets created on the client and sent as a base64-encoded bitset, where it's used as the doc value filter.

Similarly, we have the [terms lookup feature on the `terms` query](https://opensearch.org/docs/latest/query-dsl/term/terms/#terms-lookup), but it's doing a kind-of naive "fetch an array of strings" approach.

My idea is to borrow the roaring bitmap idea from the linked blog post and add that to `terms` query's lookup.

### Describe the solution you'd like

I would like to modify the terms lookup feature to add a new (opt-in) "protocol" between the main index and the term lookup index. The term lookup index should assign consistent, increasing ordinals to the values in the terms field. When the main index queries the term lookup index, it should pass a bitset of the ordinals whose values it "knows" (cached from previous requests). After finding the matching document in the term lookup index, the response should carry a bitset of matching ordinals, along with the ordinal-to-value mapping for any unknown values. This should allow us to carry very large sets of IDs across the index boundary in a compact representation.

As a next step, the term lookup should support multiple lookup keys and Boolean/bitset operations between them. The term lookup index will return the bitset after performing the Boolean operations. (The term lookup index may cache the result of these Boolean operations.)

### Related component

Search:Query Capabilities

### Describe alternatives you've considered

I drafted (on my computer) a whole idea around a new query type that would work with numeric ID bitsets, either passed in the query or stored in a custom data store (probably implemented as an OpenSearch index, but could be somewhere else). 

My concerns with that were:
1. It forces folks to use numeric IDs, which may not be a viable option.
2. It would have added lots of new APIs (to organize, create, and manage bitsets), as well as a whole new query type.

Making the existing lookup feature of `terms` query "smarter" feels like a lot less work for me and for users.

### Additional context

As I mentioned to above, I had drafted a proposal on my computer to build a whole dedicated API. While I no longer think that's the right move, my proposal did have some nice examples of possible use-cases:

#### Example 1 - Digital entitlements

An example would be a digital content entitlement system, with each document in the search index corresponding to a digital product. End-users can purchase access to content. When an end-user searches their library, they should only see content to which they have access. Updating each piece of content whenever a user makes a purchase is not practical, since a single item of content may be owned by many users. Instead, we would like each user

#### Example 2 - Multi-location retail search

A retail grocery chain offers online ordering and delivery across many store locations. They have a single catalog of products, but each location may carry a different selection of products. Product selection at any store may fluctuate as inventory sells out and new inventory is delivered. When a user in a given city searches for products, they should see items currently available from their local stores. There should be an updatable collection for each store and a user should be able to search across the union of products available from their local stores.

### PR DESCRIPTION
**Pull request by bowenlan-amzn**
### Problem

To retrieve the documents that match at least one item from a given list, we can use the [terms](https://opensearch.org/docs/latest/query-dsl/term/terms/) query. We can even save the filter list in an document and use [terms lookup](https://opensearch.org/docs/latest/query-dsl/term/terms/#terms-lookup) to fetch that and feed into the terms query.

However, as the filter becomes larger, the memory and network transmit overhead increase. And this overhead disproportionately affects the latency and TPS when the filter becomes huge (10k+ items), making it un-useable.

### Proposal

We can use [RoaringBitmap](https://github.com/RoaringBitmap/RoaringBitmap) to encode the filter which provides less memory and bandwidth usage and fast and deterministic in-memory random access or lookup.

### User Story

Users want to filter/join a main index with a bitmap filter on a numeric field. 
e.g. The index contains product ids and other data related to products. Each filter represents the owned products of a customer. The filter is a list of _numeric_ product ids.

- Users want to store the bitmap filter as documents in OpenSearch and join the filter with main index at query time. (existing terms lookup feature)
	- Users want to do boolean operations across multiple saved bitmaps at query time using the [boolean queries](https://opensearch.org/docs/2.15/query-dsl/compound/bool/).
	- (Optional) Users want to update the stored filters with bitmap operations like add, remove, etc.
- Users want to directly do terms query with a bitmap built at client side w/o having to save it.

<details>
<summary><strong>An example experience using terms lookup</strong></summary>

Uses create a RoaringBitmap for a filter on the client side, serialize the bitmap to byte array and encode the byte array using base64.

```py
bm = BitMap([111, 222, 333]) # product ids
encoded = base64.b64encode(BitMap.serialize(bm))
```

Users index/store the bitmap filter in a [binary field](https://opensearch.org/docs/latest/field-types/supported-field-types/binary/) (`customer_filter`) of an OpenSearch index (`customers`). The id of the document is the identifier of the customer associated with this filter.

```
# index mapping using binary field with stored field enabled
# we don't need it to be saved in _source so can exclude it
{
    "mappings": {
        "properties": {
            "customer_filter": {
                "type": "binary",
                "store": true
            },
            "_source": {
                "excludes": [
                    "customer_filter"
                ]
            }
        }
    }
}

POST customers/_doc/customer123
{
  "customer_filter": "OjAAAAEAAAAAAAEAEAAAAG8A3gA=" <-- base64 encoded serizlied bitmap
}
```

Users do a terms lookup query on products index (`products`) with a lookup on `customers` of certain customer id.

```
POST products/_search
{
  "query": {
    "terms": {
      "product_id": {
        "index": "customers",
        "id": "customer123",
        "path": "customer_filter",
		"store": true  <-- new parameter to do the lookup on the stored field, instead of _source
      },
      "value_type": "bitmap" <-- new parameter in terms query to specify the data type of the terms values input
    }
  }
}
```

User do a normal terms query and pass in bitmap

```
POST products/_search
{
  "query": {
    "terms": {
      "product_id": "<customer_filter>",
      "value_type": "bitmap"
    }
  }
}
```

User do a boolean query with boolean operation between multiple filters

```
POST products/_search
{
  "query": {
    "bool": {
      "must": [
        {
          "terms": {
            "product_id": "<customer1_filter>",
            "value_type": "bitmap"
          }
        }
      ],
      "should": [
        {
          "terms": {
            "product_id": "<customer2_filter>",
            "value_type": "bitmap"
          }
        },
        {
          "terms": {
            "product_id": "<customer3_filter>",
            "value_type": "bitmap"
          }
        }
      ],
      "must_not": [
        {
          "terms": {
            "product_id": "<customer4_filter>",
            "value_type": "bitmap"
          }
        }
      ]
    }
  }
}
```
</details>

### Implementation

Functional Requirements

- [x] New bitmap queries that accepts bitmap and tells what documents match with this bitmap. 2 types of queries that can be wrapped with IndexOrDocValueQuery.
	- [x] A bitmap query (Doc Value type) can efficiently verify document match
	- [x] A bitmap query (Index type) can efficiently be used to produce a lead iterator
- [x] Terms lookup on a stored bitmap in a document
	- [x] Previously terms lookup feature only fetches the `_source` of a document using get by id request. Adding support to fetch from the stored field. This is needed by binary field because its source would be the base64 encoded string.
- [x] Support bitmap as the field values in `terms` query

Non-functional Requirements

- [ ] When the filter is small, directly pass in the filter should still be better, however, when the filter becomes huge, encode and pass as bitmap would be better. Do benchmark to recommend users a general threshold here.
- [ ] The payload of a request has limit, so how big a bitmap users can query on is limited by that. Do experiment to recommend users the maximum number of numbers in a bitmap they can query on.

### Related Issues
Resolves #12341

### Check List
- [x] Functionality includes testing.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by smacrakis** [comment]
LGTM!

**Comment by bowenlan-amzn** [comment]
7/19: I have a working draft version of this bitmap filtering feature using terms lookup.
A passing yaml test is added and can give you an idea on how to use this feature

Will continue after 7/28

**Comment by bowenlan-amzn** [comment]
For some reason, the coverage of internalClusterTest is not shown in the codecov. Will need to look into that.

File: server/src/main/java/org/opensearch/index/mapper/NumberFieldMapper.java [hunk 1]
```diff
@@ -1415,6 +1445,40 @@ public static Query unsignedLongRangeQuery(
             }
             return builder.apply(l, u);
         }
+
+        static PointInSetQuery bitmapIndexQuery(String field, RoaringBitmap bitmap) {
+            final BytesRef encoded = new BytesRef(new byte[Integer.BYTES]);
+            return new PointInSetQuery(field, 1, Integer.BYTES, new PointInSetQuery.Stream() {
```
Comment (msfroh) [inline server/src/main/java/org/opensearch/index/mapper/NumberFieldMapper.java:1451]: I'm a little bit concerned about the `PointInSetQuery`'s cost estimate.

1. It iterates through all of the points, which may be a lot of points.
2. If a point is contained in a leaf of the BKD tree, it assumes that half the points/docs in the leaf (256) match the query, which is likely way too high.

I just did a quick though experiment: Suppose we have 100M docs on a shard (for simplicity, assume the shard is a single segment). The BKD tree will have 100M/512 = 200K leaves. Assuming fairly uniform distribution of documents across the available ID space, I don't think it's out of the question for a query on 1M IDs to hit every one of those leaves. As a result, the cost estimate will be 50M (when the truth is likely less than 1M).

tl;dr: The PointInSetQuery provides a good cost estimate in the general case, but probably doesn't work well for the case where your points are essentially IDs.

We'll need to think about whether this is good enough. On the plus side, it doesn't affect how things are indexed, so we have the opportunity to make it better in the future.

File: server/src/main/java/org/opensearch/search/query/BitmapDocValuesQuery.java [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+    public BitmapDocValuesQuery(String field, RoaringBitmap bitmap) {
+        this.field = field;
+        this.bitmap = bitmap;
+        if (!bitmap.isEmpty()) {
+            min = bitmap.first();
+            max = bitmap.last();
+        } else {
+            min = 0; // final field
+            max = 0;
```
Comment (msfroh) [inline server/src/main/java/org/opensearch/search/query/BitmapDocValuesQuery.java:53]: We can short-circuit the match logic by setting `max = -1`.

**Review by msfroh — APPROVED** [review]
I went through the code pretty carefully and left three comments.

In my opinion, none of them are blockers for this PR.

**Review by msfroh — APPROVED** [review]

**Comment by msfroh** [comment]
I don't seem to be able to retrigger the Mend Security Check.

It's not configured to block merges, though, so I believe it's safe to merge anyway.
